### PR TITLE
Implement bundle manifest hash checks

### PR DESCRIPTION
**Pull request by phmccarty**
The second commit in this series is the functional change. More details in the commit description.

### CODE REVIEW

File: include/swupd.h [hunk 1]
```diff
@@ -161,7 +161,8 @@ extern void apply_heuristics(struct file *file);
 
 extern int file_sort_filename(const void *a, const void *b);
 extern int file_sort_filename_reverse(const void *a, const void *b);
-extern int load_manifests(int current, int version, char *component, struct file *file, struct manifest **manifest);
+extern struct manifest *load_mom(int current, int version);
```
Comment (tpepper) [inline include/swupd.h]: "int current" here can be dropped, if...(see comment a little further down)

File: src/manifest.c [hunk 2]
```diff
@@ -574,29 +574,82 @@ static void set_untracked_manifest_files(struct manifest *manifest)
...
+ * Note that if the manifest fails to download, or if the manifest fails to be
+ * loaded into memory, this function will return NULL.
+ */
+struct manifest *load_mom(int current, int version)
 {
+	struct manifest *manifest = NULL;
 	int ret = 0;
 
-	*manifest = manifest_from_file(version, component);
+	ret = retrieve_manifests(current, version, "MoM", NULL);
```
Comment (tpepper) [inline src/manifest.c]: - ret = retrieve_manifests(current, version, "MoM", NULL);
+ ret = retrieve_manifests(version, version, "MoM", NULL);

If called in this way, then the NULL and version not less than version result in retrieve_manifests() not attempting a delta application.

Comment (tpepper) [inline src/manifest.c]: Then all the call sites become:

some_manifest_variable = load_mom(version);

That is sooooo much more readable than the current mess.  Way more sustainable / supportable / maintainable going forward.

Comment (phmccarty) [inline src/manifest.c]: Good point. That is a further useful simplication. I'll make the change.

File: src/manifest.c [hunk 3]
```diff
@@ -574,29 +574,82 @@ static void set_untracked_manifest_files(struct manifest *manifest)
...
+ * MOM manifest.
+ *
+ * Value ranges and restrictions:
+ *   - CURRENT < VERSION for 'update' subcommand
+ *   - CURRENT == VERSION for other subcommands
+ *
+ * Note that if the manifest fails to download, or if the manifest fails to be
+ * loaded into memory, this function will return NULL.
+ */
+struct manifest *load_mom(int current, int version)
```
Comment (tpepper) [inline src/manifest.c]: Add this descriptive bit here:

```
// MoM's are not huge so deltas don't give much benefit, versus bundle           
// manifests which can be huge and beg for deltas.  If there was a MoM          
// delta you'd want to validate the input MoM prior to applying a delta,        
// and then validate the output MoM.  All complicated compared to just          
// getting the whole MoM and verifying it.
```

File: src/manifest.c (hunk 2 above)
Comment (tpepper) [inline src/manifest.c]: Modulo the above little change to incrementally clean a little more aggresively, this patch looks good to me.  Really good. :)

File: src/manifest.c (hunk 3 above)
Comment (phmccarty) [inline src/manifest.c]: Okay, looks good. I'll amend the commit if there are no further comments.

**Comment by phmccarty** [comment]
New patch set uploaded.

File: src/hash.c [hunk 4]
```diff
@@ -252,3 +253,41 @@ bool verify_file(struct file *file, char *filename)
...
+		}
+
+		string_or_die(&local, "%s/%i/Manifest.%s", state_dir,
+			      current->last_change, current->filename);
+
+		if (!verify_file(bundle, local)) {
+			printf("Warning: hash check failed for Manifest.%s\n",
+			       current->filename);
+			ret = 0;
+			// free(local);
```
Comment (tpepper) [inline src/hash.c]: meant to remove?

Comment (phmccarty) [inline src/hash.c]: yes, will remove

File: src/manifest.c [hunk 5]
```diff
@@ -574,29 +574,78 @@ static void set_untracked_manifest_files(struct manifest *manifest)
...
+ * Note that if the manifest fails to download, or if the manifest fails to be
+ * loaded into memory, this function will return NULL.
+ */
+struct manifest *load_mom(int version)
 {
+	struct manifest *manifest = NULL;
 	int ret = 0;
 
-	*manifest = manifest_from_file(version, component);
+	ret = retrieve_manifests(version, version, "MoM", NULL);
```
Comment (bradTpeters) [inline src/manifest.c]: Isn't this ALWAYS going to perform a curl_get_file on the MoM, regardless of whether or not it is already on the filesystem?

**Comment by phmccarty** [comment]
Patch set no. 3 uploaded

**Comment by phmccarty** [comment]
and... patch set no. 4

File: src/hash.c [hunk 6]
```diff
@@ -252,3 +253,39 @@ bool verify_file(struct file *file, char *filename)
...
 	}
 }
+
+/* Compares the hash for BUNDLE with that listed in the Manifest.MoM.  If the
+ * hash check fails, for now, we print a warning and return success, but
+ * eventually will switch to being a fatal error and exit early.
+ */
+int verify_bundle_hash(struct manifest *manifest, struct file *bundle)
+{
+	assert(strcmp(manifest->component, "MoM") == 0);
```
Comment (tpepper) [inline src/hash.c]: one more assert here.  And the #include.

Comment (phmccarty) [inline src/hash.c]: will remove

**Comment by phmccarty** [comment]
uploaded patch set 5

**Comment by tpepper** [comment]
lgtm

File: src/hash.c [hunk 7]
```diff
@@ -252,3 +252,37 @@ bool verify_file(struct file *file, char *filename)
...
+		}
+
+		string_or_die(&local, "%s/%i/Manifest.%s", state_dir,
+			      current->last_change, current->filename);
+
+		if (!verify_file(bundle, local)) {
+			printf("Warning: hash check failed for Manifest.%s\n",
+			       current->filename);
+			ret = 0;
+		}
```
Comment (bradTpeters) [inline src/hash.c]: Do you want to return here? Else we're continuing to verify a known bad manifest

**Comment by bradTpeters** [comment]
This looks great Patrick; nicely done

**Comment by phmccarty** [comment]
Thanks, I will merge shortly.

**Pull request merged**